Multiple applications and tenants. Add EmailVerification.

bhalsey opened the pull request:
### Issue:
- https://linear.app/fusionauth/issue/ENG-2660/add-extra-data-to-snapshot (and other perf test issues)

### Problem:
- We want to simulate customers with a large set of users distributed across many applications distributed across tenants.
  - We need to know which applicationId and tenantId to use.
- We want to create email verificationIds.

### Solution:
- Add `FusionAuthBaseWorker` that looks for `numberOfTenants` and `numberOfApplications`. Given a userId it calculates the `applicationId` and `tenantId` using a 1-based index modulus operation.
- Add `FusionAuthEmailVerificationIdWorker`

### Reviewer Notes:
This is loosely based on the import scripts, but generalized so that `User-Registrations.json` can be run and continue to distribute new users across applications and tenants.

bhalsey committed:
configure for users per app per tenant

bhalsey committed:
add email-verification worker

bhalsey committed:
add FusionAuthBaseWorker

bhalsey committed:
add README notes

bhalsey committed:
Merge branch 'develop' into bhalsey/ENG-2660/multi-apps

wied03 commented on src/main/java/io/fusionauth/load/FusionAuthBaseWorker.java:
```diff
@@ -0,0 +1,95 @@
+
+  protected int applicationIndex;
+
+  protected FusionAuthClient scopedClient;
```
would it help readability to name this `tenantScopedClient` instead?

bhalsey replied:
yep, done

wied03 commented on src/main/java/io/fusionauth/load/FusionAuthBaseWorker.java:
```diff
@@ -0,0 +1,95 @@
+      if (applicationIndex == 0) {
+        applicationIndex = numberOfApplications; // indexes are 1-based
+      }
+      applicationId = applicationUUID(applicationIndex);
```
how about just calling `setApplicationIndex(applicationIndex)` here and then the method ends ?

bhalsey replied:
of course :)

wied03 reviewed (commented):
good job. just a couple things.

bhalsey committed:
PR feedback

spwitt commented on src/main/resources/User-Registrations.json:
```diff
@@ -13,7 +13,9 @@
       "factor": 1,
-      "debug": true
+      "debug": true,
+      "numberOfApplications": 1000,
```
Defaults in JSON files are 10 tenants and 100 applications. I think this should match.
```suggestion
      "numberOfApplications": 100,
```

spwitt commented on src/main/resources/Email-Verification.json:
```diff
@@ -0,0 +1,29 @@
+    "attributes": {
+      "directive": "email-verification",
+      "apiKey": "bf69486b-4733-4470-a592-f1bfce7af580",
+      "tenantId": "efb21cfc-fa60-46f4-9598-889151e58517",
```
Aren't `tenantId` and `numberOfTenants` mutually exclusive? I think we should remove this value.

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthBaseWorker.java:
```diff
@@ -0,0 +1,88 @@
+import io.fusionauth.client.FusionAuthClient;
+
+/**
+ * Worker to test creating EmailVerification Ids for users.
```
This comment is wrong. Possible suggestion.
```suggestion
 * Base worker to create deterministic Ids for FusionAuth load tests.
```

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthBaseWorker.java:
```diff
@@ -0,0 +1,88 @@
+    this.tenantScopedClient = client;
+  }
+
+  public static UUID applicationUUID(int tenantIndex) {
```
It was used correctly, but the variable was named wrong. Can we update all the methods in here to make the parameter name `index`? That avoids shadowing a class variable and makes it clearer where the value is coming from at a glance.
```suggestion
  public static UUID applicationUUID(int index) {
```

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthBaseWorker.java:
```diff
@@ -0,0 +1,88 @@
+    return new UUID(1, tenantIndex);
+  }
+
+  public static UUID tenantUUID(int tenantIndex) {
```
Are these static methods referenced outside classes that extend `FusionAuthBaseWorker`? Do they need to be passed an index when the value is a member variable?

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthCreateTenantWorker.java:
```diff
@@ -33,20 +33,20 @@
   public boolean execute() {
-    Tenant tenant = new Tenant().with(t -> t.name = "tenant_" + counter.incrementAndGet())
+    int tenantIndex = counter.incrementAndGet();
+    setTenantIndex(tenantIndex);
```
There are a lot of places in the workers and base class that have local variables with the same name as class variables. I think it's better to avoid that except in constructors. It would be easy to forget to set a value or use the wrong one. What about
```suggestion
    setTenantIndex(counter.incrementAndGet());
```
That would leave us with just one `tenantIndex` variable.

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthCreateApplicationWorker.java:
```diff
@@ -21,27 +21,32 @@
+                                               .with(a -> a.roles.add(new ApplicationRole("user")));
+
+    setApplicationIndex(applicationIndex);
+    application.tenantId = tenantId;
```
```suggestion
    setApplicationIndex(counter.incrementAndGet());

    Application application = new Application().with(a -> a.name = "application_" + applicationIndex)
                                               .with(a -> a.tenantId = tenantId)
                                               .with(a -> a.roles.add(new ApplicationRole("admin")))
                                               .with(a -> a.roles.add(new ApplicationRole("user")));
```

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthRegistrationWorker.java:
```diff
@@ -43,26 +41,30 @@ public class FusionAuthRegistrationWorker extends BaseWorker {
   @Override
   public boolean execute() {
+    int userIndex = counter.incrementAndGet();
+    setUserIndex(userIndex);
```
```suggestion
    setUserIndex(counter.incrementAndGet());
```

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthRegistrationWorker.java:
```diff
@@ -43,26 +41,30 @@ public class FusionAuthRegistrationWorker extends BaseWorker {
+    user.tenantId = tenantId;
 
-    UserRegistration userRegistration = new UserRegistration().with(r -> r.applicationId = applicationId)
+    UUID registrationAppId = applicationId != null ? applicationId : configuredApplicationId;
```
I think it's better to invert this. Use the specific `configuredApplicationId` if it was provided, otherwise fall back to the calculated value.

bhalsey replied:
updated

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthEmailVerificationIdWorker.java:
```diff
@@ -0,0 +1,55 @@
+  public boolean execute() {
+    int userIndex = counter.incrementAndGet();
+
+    setUserIndex(userIndex);
```
```suggestion
    setUserIndex(counter.incrementAndGet());
```

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthBaseWorker.java:
```diff
@@ -0,0 +1,88 @@
+    tenantScopedClient = client.setTenantId(tenantId); // this returns a new client with the tenant set
+  }
+
+  protected void setUserIndex(int userIndex) {
```
The import scripts generate all the applications for a tenant before moving on to the next one. Likewise, all the users are created for an application before moving on to the next application. These changes spread applications and users evenly as they are loaded.

I don't know that it matters much unless we want data from the two generators to be compatible.

bhalsey replied:
That was the original inspiration. The problem with that scheme, though, is how to divvy up additional registrations after creating the additional data set.

This heuristic lets us add additional users and have a predictable distribution.

spwitt commented on src/main/resources/User-Logins.json:
```diff
@@ -10,7 +10,9 @@
       "url": "https://local.fusionauth.io",
       "loginLowerBound": 1,
       "loginUpperBound": 2500,
```
Should these numbers be compatible with the user registrations created by default?

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthLoginWorker.java:
```diff
@@ -29,28 +29,29 @@
-    ClientResponse<LoginResponse, Errors> result = client.login(new LoginRequest(applicationId, email, Password));
+    int userIndex = new Random().nextInt((loginUpperBound - loginLowerBound) + 1) + loginLowerBound;
+    String email = "load_user_" + userIndex + "@fusionauth.io";
+    setUserIndex(userIndex);
```
I'd like to distinguish variable names, but I'm not sure about readability combining the random and set index lines.
```suggestion
    // Pick a random user to log in
    int index = new Random().nextInt((loginUpperBound - loginLowerBound) + 1) + loginLowerBound;
    setUserIndex(index);
    String email = "load_user_" + userIndex + "@fusionauth.io";
```

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthLoginWorker.java:
```diff
@@ -29,28 +29,29 @@
+    String email = "load_user_" + userIndex + "@fusionauth.io";
+    setUserIndex(userIndex);
+
+    UUID registrationAppId = applicationId != null ? applicationId : configuredApplicationId;
```
I think this should prefer `configuredApplicationId` if one was provided.

bhalsey replied:
updated

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthCreateTenantWorker.java:
```diff
@@ -33,20 +33,20 @@
+    Tenant tenant = new Tenant().with(t -> t.name = "tenant_" + tenantIndex)
                                 .with(t -> t.emailConfiguration.host = "localhost")
                                 .with(t -> t.emailConfiguration.port = 25)
                                 .with(t -> t.externalIdentifierConfiguration.authorizationGrantIdTimeToLiveInSeconds = 60)
```
Should this be updated to max out TTLs?

bhalsey replied:
updated the emailVerificationIdTimeToLiveInSeconds TTL. Easy enough for someone else to change if they're adding specific extIds.

spwitt reviewed (changes requested):
Mostly small cleanup suggestions. The bigger question is whether Id generation here should match `fusionauth-import-scripts` to make them compatible.

bhalsey committed:
Update src/main/java/io/fusionauth/load/FusionAuthBaseWorker.java

Co-authored-by: Spencer Witt <spencer@fusionauth.io>

bhalsey committed:
Update src/main/resources/User-Registrations.json

Co-authored-by: Spencer Witt <spencer@fusionauth.io>

bhalsey committed:
Apply suggestions from code review

Co-authored-by: Spencer Witt <spencer@fusionauth.io>

bhalsey committed:
PR feedback, separate multiple app/tenants configs to be less confusing

bhalsey committed:
update configuredApplicationId precedence. config cleanup

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthBaseWorker.java:
```diff
@@ -0,0 +1,91 @@
+    return new UUID(0, index);
+  }
+
+  protected void setApplicationIndex(int applicationIndex) {
```
I was thinking that we could rename to `index` on protected methods as well.

spwitt commented on src/main/resources/User-Logins.json:
```diff
@@ -9,7 +9,7 @@
       "url": "https://local.fusionauth.io",
       "loginLowerBound": 1,
-      "loginUpperBound": 2500,
+      "loginUpperBound": 100000,
```
Only 1000 users are created by default.

bhalsey replied:
No. 100 workers * 1000 loops creates 100_000 workers.

spwitt commented on src/main/resources/User-Logins-Multi.json:
```diff
@@ -0,0 +1,28 @@
+      "apiKey": "bf69486b-4733-4470-a592-f1bfce7af580",
+      "url": "https://local.fusionauth.io",
+      "loginLowerBound": 1,
+      "loginUpperBound": 100000,
```
Only 1000 users are created by default.

bhalsey replied:
No. 100 workers * 1000 loops creates 100_000 workers.

spwitt commented on src/main/script/set-url-host.sh:
```diff
@@ -0,0 +1,28 @@
+#
+
+# Set the url host value across all the json config files
+# Usage: ./set-url-host.sh <new_url>
```
If you call this script from it's own directory, the `find` below won't find any of the JSON files it's meant to update.

bhalsey replied:
The intention is that this is run from the `build/dist/` directory. Updated the usage and the readme to help clarify.

spwitt commented on src/main/java/io/fusionauth/load/FusionAuthBaseWorker.java:
```diff
@@ -0,0 +1,91 @@
+    this.tenantScopedClient = client;
+  }
+
+  public static UUID applicationUUID(int index) {
+    return new UUID(1, index);
+  }
```
These two static methods are not referenced outside this class, and I don't see any reason they need to be static based on their usage. If they are always called with the class member values, can we omit the parameter as well?
```suggestion
  private UUID applicationUUID() {
    return new UUID(1, applicationIndex);
  }
```
Or we could set the Ids in these methods. Or we could remove them entirely and just set the class member fields directly from the `setIndex` methods.

It's not wrong, but it bothers me that these are `public static`. I feel like that should be done when it's necessary, but generally methods should have limited visibility.

bhalsey replied:
They're simple functions, so I believe keeping them static is more appropriate. Someone might want to add multiple app registrations, so they could call this function with other randomly generated appIds. It doesn't need to be tied to class variables. I don't think limiting visibility is necessary for load testing code that is meant to be experimented with.

spwitt reviewed (commented):
Still a couple of counts in the sample files that I think should be consistent in order to be able to run using the JSON files as is.

If I'm being too bothersome with the naming and method signatures, we can move on. This is not production code.

bhalsey committed:
param name. usage help

spwitt commented on src/main/resources/User-Logins-Multi.json:
```diff
@@ -0,0 +1,28 @@
+      "loginLowerBound": 1,
+      "loginUpperBound": 100000,
+      "debug": true,
+      "numberOfApplications": 100,
```
Were registrations and logins meant to be spread across all created applications?

Based on new info about `workerCount` and `loopCount`, I think the default JSON creates 10 x 100 = 1000 applications.

bhalsey replied:
thanks. fixed.

bhalsey committed:
consistent numberOfApplications

robotdan commented on src/main/java/io/fusionauth/load/FusionAuthBaseWorker.java:
```diff
@@ -0,0 +1,91 @@
+public abstract class FusionAuthBaseWorker extends BaseWorker {
+  protected final FusionAuthClient client;
+
+  private final int numberOfApplications;
+
+  private final int numberOfTenants;
```
If you look at the load test definition, we do have several existing 'count' things - such as `loopCount`, `workerCount`, and `counter`. 

Should we name these `applicationCount` and `tenantCount` for consistency, or was that considered and and not done for a reason?

bhalsey replied:
good call. done

robotdan commented on src/main/resources/Create-Tenants.json:
```diff
@@ -1,16 +1,14 @@
 {
-  "loopCount": 100,
-  "workerCount": 100,
+  "loopCount": 10,
+  "workerCount": 1,
```
I guess it doesn't hurt to change these, but these are load test definitions. A worker count of `1` isn't much of a load test. Same for all of the `workerCount` or `loopCount` changes.

bhalsey replied:
I updated the definitions to work well as a group for what I think is the more typical use case -- a large number of users distributed across 1000 applications and 10 tenants. Open to discussion.

robotdan reviewed (commented):
**1. Copyrights**

Looks like copyrights are a bit messed up. Some new files have old dates. 


**2. Naming**

Any reason not to name the new worker definitions `-MultiTenant` or something more specific? `-Multi` could mean a lot of things. 

**3. Careful with email**

One thing to note, and perhaps this is already accounted for.. but when we use this library internally, and for that matter when customers use this library - we want to be extremely careful with email. 

In `fusionauth-app` when we run the dev mode kickstart, we create a tenant called `Load Testing` with email disabled and `verifyEmail: false`. 

We did this because it was risky to just stand up a tenant and start running load tests w/out knowing the state of the email configuration. I did it a few times by accident, and crushed our SendGrid account, and because the email addresses are fabricated, it can affect your domain deliverability score if you're not careful. 

So I just want to be sure that we - ideally not change the default config for how `fusionauth-app` dev interacts with this library. Specifically that we use the `Load Test` tenant by default which is Id `efb21cfc-fa60-46f4-9598-889151e58517`.  And that if we create more tenants, which are then used for User Registration, Login, etc., that we explicitly disable all email, and SMS to just be extra sure we don't accidentally use those services by way of running a load test. 

If a customer wants to enable those features and then use this load test framework, that is on them, but we should just be as careful as possible. 

This may all be done already, just want to mention it for clarity.

I see in the Create Tenant worker, we do use `localhost` for the email config, which was pre-existing. But in the past we did not use the results of that load test to then register new users into those tenants. 

So it appears that you will create tenants with emailing enabled (even if over localhost) and then register users. And I see you have an email verification type of workload. But if it is just to generate a verification Id, perhaps email could still be disabled? Is this only to test the generation?

bhalsey committed:
copyrights

bhalsey committed:
s/Multi/MultiTenant/

bhalsey committed:
consistent count naming

bhalsey committed:
explicit email verification false with note

bhalsey commented:
> **1. Copyrights**

Just a by product of copying files as a starter and intelliJ hiding copyright comments.

> **2. Naming**
> 
> Any reason not to name the new worker definitions `-MultiTenant` or something more specific? `-Multi` could mean a lot of things.

Updated.
 
> **3. Careful with email**

The default is verifyEmail = false. Made this explicit and added a comment.
